allennlp/models/encoder_decoders/copynet_seq2seq.py

```diff
@@ -220,7 +220,7 @@ def forward(
                     best_predictions = top_k_predictions[:, 0, :]
                     # shape: (batch_size, target_sequence_length)
                     gold_tokens = self._gather_extended_gold_tokens(
-                        target_tokens["tokens"], source_token_ids, target_token_ids
+                        target_tokens["tokens"]["tokens"], source_token_ids, target_token_ids
```

Comment:
I added `util.get_token_ids_from_text_field_tensors`, which *should* replace all places where you have to reach into the text field dictionary like this.  I wanted to leave that cleanup for a separate PR, though, as this is already pretty large, and it would require me taking more time to understand what a model is doing with these tensors and if that function really solves the problem.

In general, when a model has to reach into the dictionary like this, it's because the abstractions were used in a way they weren't designed for, and something should probably be rewritten.  I agree that it's a mess to have people actually look at content of these dictionaries (particularly for the single-tensor indexers; it's much less jarring for the multi-tensor indexers), but they were designed to hide that from you.

Reply:
Separate PR makes complete sense. Thank you!